posix: fix pthread thread specific data cleanup

msobkowski opened the pull request:
The `pthread_key_data` allocated in `pthread_setspecific()` is never deallocated. Due to that `pthread_setspecific()` is returning `ENOMEM` after using it specific amount of times, depending on size of `k_malloc` `s underlying pool.

cfriedt reviewed (commented):
Looks good - can you please also add a memory leak test (in the current test suite for pthreads keys)?

I would guess something like this should work:

* Create a pthread
* Count how many times `pthread_key_create()` and `pthread_setspecific()` can be called until getting ENOMEM.
* Terminate the thread.
* Repeat all of the above steps and ensure that the second count matches the first count.

jmichalski-ant committed:
posix: add pthread specific data deallocation test

This commit adds test to assert that thread specific data is deallocated
on thread termination.

Signed-off-by: Jakub Michalski <jmichalski@antmicro.com>

jmichalski-ant committed:
posix: fix pthread thread specific data cleanup

It was never deallocationg pthread_key_data nor it was removing the node
with it from list

Signed-off-by: Jakub Michalski <jmichalski@antmicro.com>

msobkowski commented:
Thanks for the review @cfriedt, we've added the test you suggested, could you take another look?

cfriedt commented on tests/posix/common/src/key.c:
```diff
@@ -144,6 +144,42 @@ ZTEST(key, test_correct_key_is_deleted)
+ZTEST(key, test_thread_specific_data_deallocation)
+{
+	pthread_t thread;
+	int alloc_count_t0 = 0;
+	int alloc_count_t1 = 0;
```
With hardware stack protection, and these counters in stack memory, the thread that allocates keys will likely trigger a fault on write.

I would suggest making the variables static or moving them to the global scope (or both).

```suggestion
	static int alloc_count_t0 = 0;
	static int alloc_count_t1 = 0;
```

cfriedt commented on tests/posix/common/src/key.c:
```diff
@@ -144,6 +144,42 @@ ZTEST(key, test_correct_key_is_deleted)
+		"attempt to create thread failed");
+	zassert_ok(pthread_join(thread, NULL), "failed to join thread");
+
+	zassert_ok(alloc_count_t0 != alloc_count_t1,
```
```suggestion
	zassert_equal(alloc_count_t0, alloc_count_t1,
```

cfriedt reviewed (commented):
A couple of small changes

cfriedt commented on tests/posix/common/src/key.c:
```diff
@@ -144,6 +144,42 @@ ZTEST(key, test_correct_key_is_deleted)
+	zassert_ok(pthread_create(&thread, NULL, setspecific_thread, &alloc_count_t0),
+		"attempt to create thread failed");
+	zassert_ok(pthread_join(thread, NULL), "failed to join thread");
+
```
Maybe before spawning the second thread, add a `printk()` that displays how many times a new key was allocated and assigned a value.

msobkowski commented:
@ycsin could you also take a look?

cfriedt commented on tests/posix/common/testcase.yaml:
```diff
@@ -67,6 +67,8 @@ tests:
       - CONFIG_THREAD_STACK_INFO=y
       - CONFIG_HEAP_MEM_POOL_SIZE=16384
+      - CONFIG_POSIX_THREAD_KEYS_MAX=2048
+      - CONFIG_TEST_EXTRA_STACK_SIZE=16384
```
That's a lot of extra stack size - is that necessary?
